envconfig.go

```diff
@@ -32,75 +54,227 @@ type Decoder interface {
 		}
 
-		alt := typeOfSpec.Field(i).Tag.Get("envconfig")
-		fieldName := typeOfSpec.Field(i).Name
+		alt := tof.Tag.Get("envconfig")
+		fieldName := tof.Name
 		if alt != "" {
 			fieldName = alt
 		}
 		key := strings.ToUpper(fmt.Sprintf("%s_%s", prefix, fieldName))
-		// `os.Getenv` cannot differentiate between an explicitly set empty value
-		// and an unset value. `os.LookupEnv` is preferred to `syscall.Getenv`,
-		// but it is only available in go1.5 or newer.
+		def := tof.Tag.Get("default")
+		req := tof.Tag.Get("required")
+		if req != TrueString {
+			req = FalseString
+		}
+
+		if err := process(f, tof, fieldName, key, alt, def, req); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DocumentInfo used to manage configuraiton option information for template output
+type DocumentInfo struct {
+	Key         string
+	Description string
+	Type        string
+	Default     string
+	Required    string
+}
+
+// toTypeDescription convert techie type information into something more human
+func toTypeDescription(t reflect.Type) string {
+	switch t.Kind() {
+	case reflect.Array, reflect.Slice:
+		return fmt.Sprintf("List of %s", toTypeDescription(t.Elem()))
+	case reflect.Ptr:
+		return toTypeDescription(t.Elem())
+	case reflect.String:
+		return "String"
+	case reflect.Bool:
+		return "Boolean"
+	case reflect.Int:
+		return "Integer"
+	case reflect.Int8:
+		return "Integer(8 bits)"
+	case reflect.Int16:
+		return "Integer(16 bits)"
+	case reflect.Int32:
+		return "Integer(32 bits}"
+	case reflect.Int64:
+		return "Integer(64 bits)"
+	case reflect.Uint:
+		return "Unsigned Integer"
+	case reflect.Uint8:
+		return "Unsigned Integer(8 bits)"
+	case reflect.Uint16:
+		return "Unsigned Integer(16 bits)"
+	case reflect.Uint32:
+		return "Unsigned Integer(32 bits}"
+	case reflect.Uint64:
+		return "Unsigned Integer(64 bits)"
+	case reflect.Float32:
+		return "Float"
+	case reflect.Float64:
+		return "Float(64 bits)"
+	}
+	return t.Name()
+}
+
+// Document write the default format of documentation for configuration options
+func Document(prefix string, spec interface{}, out io.Writer) error {
+	return DocumentFormat(prefix, spec, out, true, DefaultTableFormat)
+}
+
+// DocumentFormat write the configration options using the given template
+func DocumentFormat(prefix string, spec interface{}, out io.Writer, showHeader bool, format string) error {
+	var info DocumentInfo
+
+	if showHeader {
+		fmt.Fprintf(out, "USAGE: %s\n", path.Base(os.Args[0]))
+		fmt.Fprintln(out)
+		fmt.Fprintln(out, "  This application is configured via the environment. The following environment")
+		fmt.Fprintln(out, "  variables can used specified:")
+		fmt.Fprintln(out)
+	}
+
+	var tabs *tabwriter.Writer = nil
+	var tmpl *template.Template
+	var tmplSpec = format
+	var err error
+
+	// If format is prefixed with "table" then strip it off to get the per-line template, display the table headers,
+	// and inject a tab write filter
+	if strings.HasPrefix(format, "table") {
```

Comment:
There has to be a way to specify this better. Maybe in a `DocumentFormat` argument, maybe even in some `text/template` trickery like exposing a suitable [function](https://godoc.org/text/template#Template.Funcs).